### PR TITLE
Init support for filebeat multiline

### PR DESCRIPTION
**Pull request by urso**
start with multiline for filebeat (#461)

### CODE REVIEW

File: filebeat/config/config.go [hunk 1]
```diff
@@ -64,9 +64,18 @@ type HarvesterConfig struct {
 	BackoffFactor      int    `yaml:"backoff_factor"`
 	MaxBackoff         string `yaml:"max_backoff"`
 	MaxBackoffDuration time.Duration
-	ForceCloseFiles    bool     `yaml:"force_close_files"`
-	ExcludeLines       []string `yaml:"exclude_lines"`
-	IncludeLines       []string `yaml:"include_lines"`
+	ForceCloseFiles    bool             `yaml:"force_close_files"`
+	ExcludeLines       []string         `yaml:"exclude_lines"`
+	IncludeLines       []string         `yaml:"include_lines"`
+	Multiline          *MultilineConfig `yaml:"multi"`
```
Comment (tsg) [inline filebeat/config/config.go]: I'd call the option multiline to be more clear what is about.

File: filebeat/config/config.go [hunk 2]
```diff
@@ -67,9 +67,19 @@ type HarvesterConfig struct {
...
+	IncludeLines       []string         `yaml:"include_lines"`
+	Multiline          *MultilineConfig `yaml:"multiline"`
+}
+
+type MultilineConfig struct {
+	Pattern  string `yaml:"pattern"`
+	Negate   bool   `yaml:"negate"`
+	Match    string `yaml:"match"`
+	MaxLines *int   `yaml:"max_lines"`
+	MaxBytes *int   `yaml:"max_bytes"`
```
Comment (ruflin) [inline filebeat/config/config.go]: I would suggest to have max_bytes outside multiline as general configuration, as it not only applies to multiline but could also apply to a very large single line.

File: filebeat/harvester/log.go [hunk 3]
```diff
@@ -79,45 +107,46 @@ func (h *Harvester) Harvest() {
...
-
-		// Reset Backoff
-		h.backoff = h.Config.BackoffDuration
-
 		if h.shouldExportLine(text) {
-
 			// Sends text to spooler
 			event := &input.FileEvent{
-				ReadTime:     lastReadTime,
+				ReadTime:     time.Now(),
```
Comment (ruflin) [inline filebeat/harvester/log.go]: That means that it is the time when it finished reading the (multi)line and not when it started as we had before?

File: filebeat/harvester/util.go [hunk 4]
```diff
@@ -40,27 +38,18 @@ func lineEndingChars(line []byte) int {
 // readLine reads a full line into buffer and returns it.
 // In case of partial lines, readLine does return and error and en empty string
 // This could potentialy be improved / replaced by https://github.com/elastic/beats/libbeat/tree/master/common/streambuf
-func readLine(
-	reader *encoding.LineReader,
-	lastReadTime *time.Time,
```
Comment (ruflin) [inline filebeat/harvester/util.go]: Seems like lastReadTime was not used anymore already before?

File: filebeat/harvester/log.go [hunk 5]
```diff
@@ -79,45 +107,46 @@ func (h *Harvester) Harvest() {
...
-				logp.Info("Read line error: %s", err)
-				return
+			if err == errFileTruncate {
+				seeker, ok := h.file.(io.Seeker)
+				if !ok {
+					logp.Err("can not seek source")
+					return
+				}
+
+				logp.Debug("harvester", "File was truncated. Begin reading file from offset 0: %s", h.Path)
```
Comment (ruflin) [inline filebeat/harvester/log.go]: I would make this Info or Warn instead of Debug.

File: filebeat/harvester/log.go [hunk 6]
```diff
@@ -79,45 +107,46 @@ func (h *Harvester) Harvest() {
...
 
-	// XXX: lastReadTime handling last time a full line was read only?
-	//      timedReader provides timestamp some bytes have actually been read from file
-	lastReadTime := time.Now()
-
 	for {
 		// Partial lines return error and are only read on completion
-		text, bytesRead, err := readLine(reader, &timedIn.lastReadTime)
-
+		text, bytesRead, err := readLine(reader)
```
Comment (ruflin) [inline filebeat/harvester/log.go]: As far as I understand, backoff is not handled by the reader?

Comment (ruflin) [inline filebeat/harvester/log.go]: readLine returned before and error on EOF. Is this not the case anymore?

File: filebeat/harvester/log.go [hunk 7]
```diff
@@ -258,110 +271,7 @@ func (h *Harvester) initFileOffset(file *os.File) error {
...
-//
-// If error is EOF, it will check for:
-// * File truncated
-// * Older then ignore_older
-// * General file error
-//
-// If none of the above cases match, no error will be returned and file is kept open
-//
-// In case of a general error, the error itself is returned
-func (h *Harvester) handleReadlineError(lastTimeRead time.Time, err error) error {
```
Comment (ruflin) [inline filebeat/harvester/log.go]: Note for myself: This is now done in readers.go

File: filebeat/harvester/multiline.go [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+	last      []byte
+	readBytes int // bytes as read from input source
+	numLines  int
+
+	err   error // last seen error
+	state func(*multiLineReader) ([]byte, int, error)
+}
+
+const (
+	defaultMaxBytes = 10 * (1 << 20) // 10MB
```
Comment (ruflin) [inline filebeat/harvester/multiline.go]: Is that the same as LS has?

File: filebeat/harvester/readers.go [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+	errForceClose   = errors.New("file must be closed")
+	errInactive     = errors.New("file inactive")
+	errTimeout      = errors.New("timeout")
+)
+
+func newNoEOLLineReader(r lineReader) *noEOLLineReader {
+	return &noEOLLineReader{r}
+}
+
+func (n *noEOLLineReader) Next() ([]byte, int, error) {
```
Comment (ruflin) [inline filebeat/harvester/readers.go]: Add some docs here, that this line reader removes the newline chars. Perhaps it is better named noNewlineLineReader.

**Comment by ruflin** [comment]
@urso Thanks a lot for getting this started. I had a closer look, but didn't test it yet locally. Some notes:

* As it is quite a big change, it is not easy to follow all the changes. One thing I'm very caution about to change is the error handling functionality as it has (too) much magic inside. Best on a brief comparison I assume it stayed mostly the same: https://github.com/elastic/beats/pull/570/files#diff-0df022979a0f5b03f3e4f709bd9c9c4dR98
* We should definitively add system tests. I can help you here
* Interestingly I would probably have followed a different approach for the implementation. Instead of moving the functionality and error handling to the line readers, I would have kept it on a higher level. I would probably add functionality to the Harvest function. The main reason for this as I would argue the line readers should not have to deal with timeouts, backoffs or the error handling. Looking at the implementation it could be that it is more a "naming" discussion, as the basic line by line reader still exist. This is probably easier to discuss via chat with a small diagram :-)

File: filebeat/config/config.go (hunk 2 above)
Comment (urso) [inline filebeat/config/config.go]: Good point. Will do. same for max_bytes.

File: filebeat/harvester/log.go (hunk 3 above)
Comment (urso) [inline filebeat/harvester/log.go]: lastReadTime was set to time.Now() right before creating the event: https://github.com/elastic/beats/blob/master/filebeat/harvester/log.go#L111

I just removed the variable. We can update the lineReader interface (or create another interface for last line-reading stage) to return timestamps and include the timestamp in the state-machine.

File: filebeat/harvester/util.go (hunk 4 above)
Comment (urso) [inline filebeat/harvester/util.go]: nope, was not used anymore, due to us removing partial line support

File: filebeat/harvester/log.go (hunk 5 above)
Comment (urso) [inline filebeat/harvester/log.go]: copy'n paste, but will change to Info.

File: filebeat/harvester/log.go (hunk 6 above)
Comment (urso) [inline filebeat/harvester/log.go]: with us removing partial line support, readLine did not return before anymore. Retrying + backoff is moved to first stage in reader (which more resembles mod-time logic in prospector) to not confuse multiline reader state machine.

File: filebeat/harvester/multiline.go (hunk 8 above)
Comment (urso) [inline filebeat/harvester/multiline.go]: yes

File: filebeat/harvester/readers.go (hunk 9 above)
Comment (urso) [inline filebeat/harvester/readers.go]: noNewlineLineReader sounds funny. Yeah, missing some good naming here.

**Comment by urso** [comment]
Error handling did not change. But I found error handling or retry functionallty to be too much inter winded. I did try to build "small" composable units being responsible for just one functionality when reading lines. I did find there is no 'higher' level without abstraction leak or doing all multiline-handling much more inter winded with error handling.
So I pushed retry logic + backoff to an earlier stage so not to confuse the multiline-reader state machine. I didn't want to handle multiline in main-loop, which is already hard to follow.

Processing stages (pipe-line is setup as follows):

    file input (io.Reader) ->
    logFileReader (io.Reader with retry and backoff) ->
    encoding.LineReader (read lines and apply codec) ->
    [optional] timeoutReader (force multiline Reader to publish current state) ->
    [optional] multilineReader (state-machine combining multiple lines) ->
    noEOLLineReader (remove trailing newline) ->
    harvester loop (publish line event and quit loop on error)

The logFileReader implements error handling on lower level, so state machines in LineReader and multilineReader have sane error handling + timeout reader can be safely shutdown if error is not nil. Having the logFileReader removes the requirement of readers being still valid after EOF.

**Comment by urso** [comment]
In this PR I've implemented a layered approach:

1. raw input layer
2. processing layer (codec, line reader, multiline, remove trailing newline)
3. event publisher loop

Layers 1 and 2 use the io.Reader interface and layers 2 and 3 use the lineReader interface. By swapping the input layer, all processing layers can be reused for different inputs (e.g. TCP/unix socket or stdin). I think layers for dealing with complexity + composing processing elements by interfaces is a quite powerful and easy to use mechanism that fits well with i/o interfaces in go stdlib. Some more refactoring + cleanup in the harvester is required in general.

It is the input layer which must behave different for different source types. For log files we want backoff + retry on EOF, but not for sockets or stdin. It's up to the input layer to push bytes until some well-defined EOF is reached. For log files the input layer is implemented by `logFileReader`. Sockets and file handle for stdin can be used as is (we don't want to ignore EOF). The error handling as it has been implemented in log.go was a good fit for log files, but not so much for stdin for example (which it was used for).

There is no real error handling in the line readers, but I have had to move some error handling to the input layer, so statefull processors in layer 2 are not affected by error handling/recovery in layer 3.

I think another advantage of having the logFileReader handling the reading from files is, we're more flexible in adding file reading features. For example:
- on windows close the file after EOF and retry (by re-opening the file) after backoff seconds. No state in upper layers will be affected
- have central authority watching number of open files and schedule temporary closing/opening of files in logFileReader in order to limit the total number of file descriptors used by filebeat.

File: filebeat/config/config.go (hunk 2 above)
Comment (ruflin) [inline filebeat/config/config.go]: What you mean by same for max_bytes?

File: filebeat/harvester/log.go (hunk 3 above)
Comment (ruflin) [inline filebeat/harvester/log.go]: :+1:

File: filebeat/config/config.go (hunk 2 above)
Comment (urso) [inline filebeat/config/config.go]: I meant to say max_lines... but max_lines makes no sense. Just will do for max_bytes. Sorry for the noise.

File: filebeat/harvester/log.go (hunk 3 above)
Comment (urso) [inline filebeat/harvester/log.go]: done

File: filebeat/harvester/log.go [hunk 10]
```diff
@@ -47,9 +50,42 @@ func NewHarvester(
 	return h, nil
 }
 
+func createLineReader(
+	in FileSource,
+	codec encoding.Encoding,
+	bufferSize int,
+	maxBytes int,
+	readerConfig logFileReaderConfig,
```
Comment (ruflin) [inline filebeat/harvester/log.go]: Should be maxBytes part of logFileReaderConfig? Same for bufferSize and codec?

File: filebeat/harvester/log.go [hunk 11]
```diff
@@ -47,9 +50,42 @@ func NewHarvester(
...
+		p, err = processor.NewMultiline(p, maxBytes, mlrConfig)
+		if err != nil {
+			return nil, err
+		}
+
+		return processor.NewStripNewline(p), nil
+	}
+
+	p = processor.NewStripNewline(p)
+	return processor.NewLimitProcessor(p, maxBytes), nil
```
Comment (ruflin) [inline filebeat/harvester/log.go]: You add here Processor to the name but we don't do it for the other processors. Should we call it just NewLimit(...)?

File: filebeat/harvester/log.go [hunk 12]
```diff
@@ -47,9 +50,42 @@ func NewHarvester(
...
+) (processor.LineProcessor, error) {
+	var p processor.LineProcessor
+	var err error
+
+	fileReader, err := newLogFileReader(in, readerConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err = processor.NewLineSource(fileReader, codec, bufferSize)
```
Comment (ruflin) [inline filebeat/harvester/log.go]: That is the only name I'm not happy yet. What is this processor on exactly doing?

File: filebeat/harvester/log.go (hunk 10 above)
Comment (urso) [inline filebeat/harvester/log.go]: logFileReadConfig is the input layer config (file reading only). The other options codec, bufferSize and maxBytes are used to configure the line processing layer.

File: filebeat/harvester/log.go (hunk 11 above)
Comment (urso) [inline filebeat/harvester/log.go]: Used to have NewLimit first, but found it to be weird.

File: filebeat/harvester/processor/timeout.go [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+	return &timeoutProcessor{
+		reader:  in,
+		signal:  signal,
+		timeout: timeout,
+		ch:      make(chan lineMessage, 1),
+	}
+}
+
+func (tr *timeoutProcessor) Next() (Line, error) {
```
Comment (ruflin) [inline filebeat/harvester/processor/timeout.go]: the variable name does not match anymore the type (tr)

File: filebeat/harvester/log.go (hunk 12 above)
Comment (urso) [inline filebeat/harvester/log.go]: it converts the encoding.LineReader (the source) into a LineProcessor adding the timestamps. Kind of an adapter.

File: filebeat/harvester/readers.go [hunk 14]
```diff
@@ -0,0 +1,166 @@
+package harvester
```
Comment (ruflin) [inline filebeat/harvester/readers.go]: It as probably now reader.go as it only contains one reader?

Comment (urso) [inline filebeat/harvester/readers.go]: hehe

File: filebeat/harvester/log.go (hunk 12 above)
Comment (ruflin) [inline filebeat/harvester/log.go]: NewReaderConverter?

Comment (urso) [inline filebeat/harvester/log.go]: it's not exactly a converter, as it instantiates the encoder and line reader from a simple io.Reader:

https://github.com/elastic/beats/pull/570/files#diff-af17398a629e85894cc774256ab08a32R39

File: filebeat/docs/configuration.asciidoc [hunk 15]
```diff
@@ -148,6 +148,50 @@ document. The default value is `log`.
...
+-------------------------------------------------------------------------------------
+multiline:
+    pattern: ^\[
+    match: after
+    negate: true
+-------------------------------------------------------------------------------------
+
+====== pattern
+
+The regexp Pattern that hast to be matched. The example pattern matches all lines starting with [
```
Comment (ruflin) [inline filebeat/docs/configuration.asciidoc]: Typo: s/hast/has/

File: filebeat/etc/beat.yml [hunk 16]
```diff
@@ -32,11 +32,11 @@ filebeat:
...
       # matching any regular expression from the list. The include_lines is called before
-      # exclude_lines. By default, no lines are dropped. 
+      # exclude_lines. By default, no lines are dropped.
       # exclude_lines: ["^DBG"]
 
       # Include lines. A list of regular expressions to match. It exports the lines that are
-      # matching any regular expression from the list. The include_lines is called before 
+      # matching any regular expression from the list. The include_lines is called before
       # exclude_lines. By default, all the lines are exported.
       # include_lines: ["^ERR", "^WARN"]
```
Comment (ruflin) [inline filebeat/etc/beat.yml]: Not related, but we should remove here and for exclude_lines the space in front.

**Comment by ruflin** [comment]
LGTM

File: filebeat/docs/configuration.asciidoc (hunk 15 above)
Comment (andrewkroh) [inline filebeat/docs/configuration.asciidoc]: I think `Pattern` can be lower-cased here. Do you think we should say that it is the same syntax as Perl or link to https://github.com/google/re2/wiki/Syntax.

File: filebeat/docs/configuration.asciidoc [hunk 17]
```diff
@@ -148,6 +148,50 @@ document. The default value is `log`.
...
+
+====== negate
+
+Defines if the pattern set under pattern should be negated or not. Default is false.
+
+====== match
+
+Match can be set to "after" or "before". It is used to define if lines should be append to a pattern
```
Comment (andrewkroh) [inline filebeat/docs/configuration.asciidoc]: `s/append/appended/`

File: filebeat/docs/configuration.asciidoc [hunk 18]
```diff
@@ -148,6 +148,50 @@ document. The default value is `log`.
...
+
+====== pattern
+
+The regexp Pattern that hast to be matched. The example pattern matches all lines starting with [
+
+====== negate
+
+Defines if the pattern set under pattern should be negated or not. Default is false.
+
+====== match
```
Comment (andrewkroh) [inline filebeat/docs/configuration.asciidoc]: `match` is required, right?

File: filebeat/etc/beat.yml [hunk 19]
```diff
@@ -73,6 +73,35 @@ filebeat:
...
+      # Maximum number of bytes a single log event can have
+      # All bytes after max_bytes are discarded and not sent. The default is 10MB.
+      # This is especially useful for multiline log messages which can get large.
+      #max_bytes: 10485760
+
+      # Mutiline can be used for log messages spanning multiple lines. This is common
+      # for Java Stack Traces or C-Line Continuation
+      #multiline:
+
+        # The regexp Pattern that hast to be matched. The example pattern matches all lines starting with [
```
Comment (andrewkroh) [inline filebeat/etc/beat.yml]: `s/hast/has`

File: filebeat/docs/configuration.asciidoc (hunk 15 above)
Comment (urso) [inline filebeat/docs/configuration.asciidoc]: we're using [CompilePOSIX](https://golang.org/pkg/regexp/#CompilePOSIX) to compile the regular expressions. These are 'mostly' compatible to egrep + some quirks (see documentation).

TBH I find it quit difficult to understand the impact on syntax when given this line in documentation:

>  restricts the regular expression to POSIX ERE (egrep) syntax

Comment (urso) [inline filebeat/docs/configuration.asciidoc]: even worse, you have to keep YAML string escaping in mind when doing the regular expression

File: filebeat/docs/configuration.asciidoc (hunk 18 above)
Comment (urso) [inline filebeat/docs/configuration.asciidoc]: yes, match and pattern is required.

File: filebeat/docs/configuration.asciidoc (hunk 15 above)
Comment (ruflin) [inline filebeat/docs/configuration.asciidoc]: That also relates the the Regexp expression for exclude and include, right? If yes, I suggest to create a separate page with docs to regexp we can link to.

Comment (urso) [inline filebeat/docs/configuration.asciidoc]: Yes, we use same compile function.

Comment (ruflin) [inline filebeat/docs/configuration.asciidoc]: Should we open an issue for that to not forget about it and discuss it with @dedemorton ?
